src/arithmetic/fields.rs

```diff
@@ -174,6 +211,197 @@ pub trait FieldExt:
     }
 }
 
+/// Parameters for a perfect hash function used in square root computation.
+#[derive(Debug)]
+struct SqrtHasher<F: FieldExt> {
+    hash_xor: u32,
+    hash_mod: usize,
+    marker: PhantomData<F>,
+}
+
+impl<F: FieldExt> SqrtHasher<F> {
+    /// Returns a perfect hash of x for use with SqrtTables::inv.
+    fn hash(&self, x: &F) -> usize {
+        ((x.get_lower_32() ^ self.hash_xor) as usize) % self.hash_mod
+    }
+}
+
+/// Tables used for square root computation.
+#[derive(Debug)]
+pub struct SqrtTables<F: FieldExt> {
+    hasher: SqrtHasher<F>,
+    inv: Vec<u8>,
+    g0: [F; 256],
+    g1: [F; 256],
+    g2: [F; 256],
+    g3: [F; 129],
```

Comment:
These should be at least `Box<[F; 256]>` to save on stack space; this looks like it'll consume 26 KB.